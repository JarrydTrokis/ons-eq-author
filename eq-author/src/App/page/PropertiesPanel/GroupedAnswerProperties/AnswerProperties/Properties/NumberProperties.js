import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import { flatMap } from "lodash";
import { groupBy, getOr } from "lodash/fp";
import { useMutation } from "@apollo/react-hooks";

import { colors } from "constants/theme";
import { UNIT } from "constants/answer-types";
import {
  characterErrors,
  SELECTION_REQUIRED,
} from "constants/validationMessages";
import { unitConversion } from "constants/unit-types";

import { Column } from "components/Grid";
import Collapsible from "components/Collapsible";
import IconText from "components/IconText";
import { Autocomplete } from "components/Autocomplete";

import { ToggleProperty } from "../Properties";

import ValidationErrorIcon from "../../validation-warning-icon.svg?inline";
import InlineField from "../../InlineField";
import MultiLineField from "../../MultiLineField";
import Decimal from "../../Decimal";

import AnswerValidation from "App/page/Design/Validation/AnswerValidation";

import updateAnswersOfTypeMutation from "graphql/updateAnswersOfType.graphql";

const ValidationWarning = styled(IconText)`
  color: ${colors.red};
  margin-top: 0.5em;
  justify-content: normal;
`;

const ValidationWarningUnit = styled(ValidationWarning)`
  margin-top: -0.5em;
`;

const filterCondition = (x, query) =>
  x.unit.toLowerCase().includes(query.toLowerCase().trim()) ||
  x.abbreviation.toLowerCase().includes(query.toLowerCase().trim()) ||
  x.type.toLowerCase().includes(query.toLowerCase().trim()) ||
  query.toLowerCase().trim().startsWith(x.unit.toLowerCase()) ||
  query.toLowerCase().trim().startsWith(x.abbreviation.toLowerCase());

const filterUnitOptions = (options, query) => {
  const common = Object.values(options)
    .filter((x) => filterCondition(x, query))
    .map((option, index) => (
      <span key={`unit-option-${index}`} value={option.unit}>
        {option.unit} <span aria-hidden="true">({option.abbreviation})</span>
      </span>
    ));
  if (!query.length) {
    const categorized = flatMap(
      groupBy("type", unitConversion),
      (item, index) => [index, ...item]
    );
    const categories = categorized.map((option, index) =>
      typeof option === "string" ? (
        <span category="true">{option}</span>
      ) : (
        <span key={`unit-option-${index}`} value={option.unit}>
          {option.unit} <span aria-hidden="true">({option.abbreviation})</span>
        </span>
      )
    );
    return [common, categories];
  }

  return [common];
};

const NumberProperties = ({
  answer,
  hasDecimalInconsistency,
  handleChange,
  page,
  getId,
}) => {
  const id = getId("required", answer.id);

  const hasUnitError = getOr([], "validationErrorInfo.errors", page)
    .map(({ field }) => field)
    .includes("unit");

  const handleUnitChange = (type, properties) => {
    updateAnswersOfType({
      variables: { input: { type, questionPageId: page.id, properties } },
    });
  };

  const [updateAnswersOfType] = useMutation(updateAnswersOfTypeMutation);

  return (
    <Collapsible
      title={`${answer.type} properties`}
      withoutHideThis
      variant="content"
    >
      <Column cols={3} gutters>
        <InlineField id={id} label={"Required"}>
          <ToggleProperty
            data-test="answer-properties-required-toggle"
            id={id}
            onChange={handleChange}
            value={answer.properties.required}
          />
        </InlineField>
      </Column>
      <Column cols={6} gutters={false}>
        <InlineField id={id} label={"Decimals"}>
          <Decimal
            id={id}
            data-test="decimals"
            onBlur={(decimals) => {
              handleChange(answer.type, {
                decimals,
              });
            }}
            value={answer.properties.decimals}
            hasDecimalInconsistency={hasDecimalInconsistency}
          />
        </InlineField>
        {answer.type === UNIT && (
          <>
            <MultiLineField id="unit" label={"Type"}>
              <Autocomplete
                options={unitConversion}
                filter={filterUnitOptions}
                placeholder={"Select a unit type"}
                updateOption={(element) => {
                  handleUnitChange(answer.type, {
                    unit: element && element.children[0]?.getAttribute("value"),
                  });
                }}
                hasError={hasUnitError}
                defaultValue={
                  answer.properties.unit
                    ? `${answer.properties.unit} (${
                        unitConversion[answer.properties.unit].abbreviation
                      })`
                    : ""
                }
              />
            </MultiLineField>
            {hasUnitError && (
              <ValidationWarningUnit
                icon={ValidationErrorIcon}
                data-test="unitRequired"
              >
                {SELECTION_REQUIRED}
              </ValidationWarningUnit>
            )}
          </>
        )}
      </Column>
      {hasDecimalInconsistency && (
        <ValidationWarning icon={ValidationErrorIcon}>
          {characterErrors.DECIMAL_MUST_BE_SAME}
        </ValidationWarning>
      )}
      <AnswerValidation answer={answer} />
    </Collapsible>
  );
};

NumberProperties.propTypes = {
  hasDecimalInconsistency: PropTypes.bool,
  handleChange: PropTypes.func,
  answer: PropTypes.object, //eslint-disable-line
  page: PropTypes.object, //eslint-disable-line
  getId: PropTypes.func,
};

export default NumberProperties;
