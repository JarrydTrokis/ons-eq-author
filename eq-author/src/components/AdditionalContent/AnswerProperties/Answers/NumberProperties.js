import React from "react";
import PropTypes from "prop-types";

import styled from "styled-components";
import Decimal from "components/AdditionalContent/AnswerProperties/Decimal";
import Required from "components/AdditionalContent/Required";
import MultiLineField from "components/AdditionalContent/AnswerProperties/Format/MultiLineField";

const Container = styled.div`
  display: flex;
  align-items: center;
`;

const NumberProperties = ({ answer, onUpdateRequired, onUpdateDecimal }) => {
  return (
    <>
      <Container>
        <MultiLineField id="decimal" htmlFor="decimal" label="Decimal places">
          <Decimal
            id={answer.id}
            answer={answer}
            data-test="decimals"
            onBlur={onUpdateDecimal}
            value={answer.properties.decimals}
            hasDecimalInconsistency={false}
          />
        </MultiLineField>
      </Container>
      <Container>
        <Required answer={answer} onChange={onUpdateRequired} />
      </Container>
    </>
  );
};

NumberProperties.propTypes = {
  onUpdateRequired: PropTypes.func,
  onUpdateDecimal: PropTypes.func,
  answer: PropTypes.object, //eslint-disable-line
};

export default NumberProperties;
