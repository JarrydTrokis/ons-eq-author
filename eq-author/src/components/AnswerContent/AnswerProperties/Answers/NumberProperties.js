import React from "react";
import PropTypes from "prop-types";

import styled from "styled-components";
import Decimal from "components/AnswerContent/AnswerProperties/Decimal";
import Required from "components/AnswerContent/Required";
import MultiLineField from "components/AnswerContent/AnswerProperties/Format/MultiLineField";

const Container = styled.div`
  display: flex;
  align-items: center;
`;

const NumberProperties = ({
  answer,
  page,
  updateAnswer,
  updateAnswerOfType,
}) => {
  return (
    <>
      <Container>
        <MultiLineField id="decimal" htmlFor="decimal" label="Decimal places">
          <Decimal
            id={answer.id}
            answer={answer}
            page={page}
            data-test="decimals"
            updateAnswerOfType={updateAnswerOfType}
            value={answer.properties.decimals}
          />
        </MultiLineField>
      </Container>
      <Container>
        <Required answer={answer} updateAnswer={updateAnswer} />
      </Container>
    </>
  );
};

NumberProperties.propTypes = {
  answer: PropTypes.object, //eslint-disable-line
  page: PropTypes.object, //eslint-disable-line
  updateAnswer: PropTypes.func,
  updateAnswerOfType: PropTypes.func,
};

export default NumberProperties;
