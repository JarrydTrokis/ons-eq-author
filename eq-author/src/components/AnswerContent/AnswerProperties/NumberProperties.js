import React from "react";
import PropTypes from "prop-types";
import Decimal from "components/AnswerContent/Decimal";
import Required from "components/AnswerContent/Required";
import MultiLineField from "components/AnswerContent/Format/MultiLineField";

const NumberProperties = ({
  answer,
  page,
  updateAnswer,
  updateAnswerOfType,
}) => {
  return (
    <>
      <MultiLineField id="decimals" label="Decimal places">
        <Decimal
          id="decimals"
          answer={answer}
          page={page}
          data-test="decimals"
          updateAnswerOfType={updateAnswerOfType}
          value={answer.properties.decimals}
        />
      </MultiLineField>
      <Required answer={answer} updateAnswer={updateAnswer} />
    </>
  );
};

NumberProperties.propTypes = {
  answer: PropTypes.object, //eslint-disable-line
  page: PropTypes.object, //eslint-disable-line
  updateAnswer: PropTypes.func,
  updateAnswerOfType: PropTypes.func,
};

export default NumberProperties;
