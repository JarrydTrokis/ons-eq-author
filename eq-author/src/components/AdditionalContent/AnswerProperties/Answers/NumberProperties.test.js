import React from "react";
import { shallow } from "enzyme";

import { useMutation } from "@apollo/react-hooks";

import NumberProperties from "./NumberProperties";

const createWrapper = (props = {}, render = shallow) => {
  return render(<NumberProperties {...props} />);
};

jest.mock("@apollo/react-hooks", () => ({
  useMutation: jest.fn(),
}));
useMutation.mockImplementation(jest.fn(() => [jest.fn()]));

describe("Required Property", () => {
  let props, wrapper;

  beforeEach(() => {
    props = {
      answer: {
        id: "1",
        type: "number",
        properties: {
          decimals: 0,
        },
        validationErrorInfo: {
          totalCount: 0,
          errors: [],
          id: "1",
        },
      },
      onChange: jest.fn(),
      getId: jest.fn(),
    };

    wrapper = createWrapper(props, shallow);
  });

  it("should render", () => {
    const updateAnswersOfType = jest.fn();
    useMutation.mockImplementation(() => [updateAnswersOfType]);
    expect(wrapper).toMatchSnapshot();
  });
});
