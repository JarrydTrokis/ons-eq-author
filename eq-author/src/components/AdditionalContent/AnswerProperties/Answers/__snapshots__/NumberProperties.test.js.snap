// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Required Property should render 1`] = `
<Collapsible
  errorCount={0}
  title="number properties"
  variant="properties"
  withoutHideThis={true}
>
  <NumberProperties__Container>
    <Required
      answer={
        Object {
          "id": "1",
          "properties": Object {
            "decimals": 0,
          },
          "type": "number",
          "validationErrorInfo": Object {
            "errors": Array [],
            "id": "1",
            "totalCount": 0,
          },
        }
      }
      getId={
        [MockFunction] {
          "calls": Array [
            Array [
              "required",
              "1",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      label="Answer required"
    />
    <NumberProperties__VerticalRule />
    <InlineField
      label="Decimals"
    >
      <Decimal
        answer={
          Object {
            "id": "1",
            "properties": Object {
              "decimals": 0,
            },
            "type": "number",
            "validationErrorInfo": Object {
              "errors": Array [],
              "id": "1",
              "totalCount": 0,
            },
          }
        }
        data-test="decimals"
        id="1"
        value={0}
      />
    </InlineField>
  </NumberProperties__Container>
  <NumberProperties__HorizontalRule />
  <Label
    bold={true}
  >
    Validation settings
  </Label>
  <Memo(AnswerValidation)
    answer={
      Object {
        "id": "1",
        "properties": Object {
          "decimals": 0,
        },
        "type": "number",
        "validationErrorInfo": Object {
          "errors": Array [],
          "id": "1",
          "totalCount": 0,
        },
      }
    }
  />
</Collapsible>
`;
